File: server/app/models/pet.py
"""
Pet data models and schemas
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
from enum import Enum

class PetSpecies(str, Enum):
    """Pet species enumeration"""
    DOG = "dog"
    CAT = "cat"

class PetActivityLevel(str, Enum):
    """Pet activity level enumeration"""
    LOW = "low"
    MODERATE = "moderate"
    HIGH = "high"

class PetLifeStage(str, Enum):
    """Pet life stage enumeration"""
    PUPPY = "puppy"
    ADULT = "adult"
    SENIOR = "senior"
    PREGNANT = "pregnant"
    LACTATING = "lactating"

class PetBreed(BaseModel):
    """Pet breed information"""
    name: str
    species: PetSpecies

class PetBase(BaseModel):
    """Base pet model with common fields"""
    name: str = Field(..., min_length=1, max_length=100)
    species: PetSpecies
    breed: Optional[str] = None
    birthday: Optional[date] = None
    weight_kg: Optional[float] = Field(None, ge=0.1, le=200.0)
    activity_level: Optional[PetActivityLevel] = None
    image_url: Optional[str] = None
    known_sensitivities: List[str] = Field(default_factory=list)
    vet_name: Optional[str] = None
    vet_phone: Optional[str] = None

class PetCreate(PetBase):
    """Pet creation model"""
    pass

class PetUpdate(BaseModel):
    """Pet update model"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    breed: Optional[str] = None
    birthday: Optional[date] = None
    weight_kg: Optional[float] = Field(None, ge=0.1, le=200.0)
    activity_level: Optional[PetActivityLevel] = None
    image_url: Optional[str] = None
    known_sensitivities: Optional[List[str]] = None
    vet_name: Optional[str] = None
    vet_phone: Optional[str] = None

class PetResponse(PetBase):
    """Pet response model"""
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime
    
    @property
    def age_months(self) -> Optional[int]:
        """Calculate age in months from birthday"""
        if not self.birthday:
            return None
        
        today = date.today()
        years = today.year - self.birthday.year
        months = today.month - self.birthday.month
        
        if today.day < self.birthday.day:
            months -= 1
        
        if months < 0:
            years -= 1
            months += 12
        
        return years * 12 + months
    
    @property
    def age_years(self) -> Optional[float]:
        """Calculate age in years from birthday"""
        if not self.birthday:
            return None
        
        today = date.today()
        years = today.year - self.birthday.year
        months = today.month - self.birthday.month
        
        if today.day < self.birthday.day:
            months -= 1
        
        if months < 0:
            years -= 1
            months += 12
        
        return years + (months / 12.0)
    
    @property
    def life_stage(self) -> PetLifeStage:
        """Calculate life stage based on age and species"""
        age_months = self.age_months
        if not age_months:
            return PetLifeStage.ADULT
        
        if age_months < 12:
            return PetLifeStage.PUPPY
        elif age_months >= 84:  # 7 years
            return PetLifeStage.SENIOR
        else:
            return PetLifeStage.ADULT
    
    @property
    def effective_activity_level(self) -> PetActivityLevel:
        """Get effective activity level (defaults to moderate if not set)"""
        return self.activity_level or PetActivityLevel.MODERATE
    
    class Config:
        from_attributes = True
        # Exclude computed properties from JSON serialization
        exclude = {'age_months', 'age_years'}

File: pet-allergy-scanner/SniffTest/Features/Pets/Models/Pet.swift
//
//  Pet.swift
//  SniffTest
//
//  Created by Steven Matos on 9/26/25.
//

import Foundation

/// Pet data model representing a pet profile
struct Pet: Codable, Identifiable, Equatable, Hashable {
    let id: String
    let userId: String
    let name: String
    let species: PetSpecies
    let breed: String?
    let birthday: Date?
    var weightKg: Double?
    let activityLevel: PetActivityLevel?
    let imageUrl: String?
    let knownSensitivities: [String]
    let vetName: String?
    let vetPhone: String?
    let createdAt: Date
    let updatedAt: Date
    
    /// Validation for pet data
    var isValid: Bool {
        return !id.isEmpty && !userId.isEmpty && !name.isEmpty && name.count >= 2
    }
    
    /// Display name for the pet
    var displayName: String {
        if let breed = breed, !breed.isEmpty {
            return "\(name) (\(breed))"
        }
        return name
    }
    
    /// Calculate age in months from birthday
    var ageMonths: Int? {
        guard let birthday = birthday else { return nil }
        
        let calendar = Calendar.current
        let now = Date()
        let components = calendar.dateComponents([.year, .month], from: birthday, to: now)
        
        guard let years = components.year, let months = components.month else { return nil }
        return years * 12 + months
    }
    
    /// Calculate age in years from birthday
    var ageYears: Double? {
        guard let birthday = birthday else { return nil }
        
        let calendar = Calendar.current
        let now = Date()
        let components = calendar.dateComponents([.year, .month], from: birthday, to: now)
        
        guard let years = components.year, let months = components.month else { return nil }
        return Double(years) + (Double(months) / 12.0)
    }
    
    /// Age description in human-readable format
    var ageDescription: String? {
        guard let ageMonths = ageMonths else { return nil }
        
        if ageMonths < 12 {
            return "\(ageMonths) month\(ageMonths == 1 ? "" : "s") old"
        } else {
            let years = ageMonths / 12
            let remainingMonths = ageMonths % 12
            if remainingMonths == 0 {
                return "\(years) year\(years == 1 ? "" : "s") old"
            } else {
                return "\(years) year\(years == 1 ? "" : "s"), \(remainingMonths) month\(remainingMonths == 1 ? "" : "s") old"
            }
        }
    }
    
    /// Calculate life stage based on age and species
    var lifeStage: PetLifeStage {
        guard let ageMonths = ageMonths else { return .adult }
        
        switch species {
        case .dog:
            if ageMonths < 12 {
                return .puppy
            } else if ageMonths >= 84 { // 7 years for dogs
                return .senior
            } else {
                return .adult
            }
        case .cat:
            if ageMonths < 12 {
                return .puppy
            } else if ageMonths >= 84 { // 7 years for cats
                return .senior
            } else {
                return .adult
            }
        }
    }
    
    /// Get default activity level if not set
    var effectiveActivityLevel: PetActivityLevel {
        return activityLevel ?? .moderate
    }
    
    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case name
        case species
        case breed
        case birthday
        case weightKg = "weight_kg"
        case activityLevel = "activity_level"
        case imageUrl = "image_url"
        case knownSensitivities = "known_sensitivities"
        case vetName = "vet_name"
        case vetPhone = "vet_phone"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

/// Pet species enumeration
enum PetSpecies: String, Codable, CaseIterable {
    case dog = "dog"
    case cat = "cat"
    
    var displayName: String {
        switch self {
        case .dog:
            return "Dog"
        case .cat:
            return "Cat"
        }
    }
    
    var icon: String {
        switch self {
        case .dog:
            return "dog.fill"
        case .cat:
            return "cat.fill"
        }
    }
}

/// Pet activity level enumeration
enum PetActivityLevel: String, Codable, CaseIterable {
    case low = "low"
    case moderate = "moderate"
    case high = "high"
    
    var displayName: String {
        switch self {
        case .low:
            return "Low Activity"
        case .moderate:
            return "Moderate Activity"
        case .high:
            return "High Activity"
        }
    }
    
    var description: String {
        switch self {
        case .low:
            return "Indoor, minimal exercise"
        case .moderate:
            return "Regular walks, some play"
        case .high:
            return "Very active, lots of exercise"
        }
    }
}

/// Pet life stage enumeration
enum PetLifeStage: String, Codable, CaseIterable {
    case puppy = "puppy"
    case adult = "adult"
    case senior = "senior"
    case pregnant = "pregnant"
    case lactating = "lactating"
    
    var displayName: String {
        switch self {
        case .puppy:
            return "Puppy/Kitten"
        case .adult:
            return "Adult"
        case .senior:
            return "Senior"
        case .pregnant:
            return "Pregnant"
        case .lactating:
            return "Lactating"
        }
    }
}

/// Pet creation model for new pet profiles
struct PetCreate: Codable {
    let name: String
    let species: PetSpecies
    let breed: String?
    let birthday: Date?
    let weightKg: Double?
    let activityLevel: PetActivityLevel?
    let imageUrl: String?
    let knownSensitivities: [String]
    let vetName: String?
    let vetPhone: String?
    
    /// Validation for pet creation data
    var isValid: Bool {
        return !name.isEmpty && name.count >= 2 && name.count <= 50
    }
    
    /// Validation errors for pet creation
    var validationErrors: [String] {
        var errors: [String] = []
        
        if name.isEmpty {
            errors.append("Pet name is required")
        } else if name.count < 2 {
            errors.append("Pet name must be at least 2 characters")
        } else if name.count > 50 {
            errors.append("Pet name must be less than 50 characters")
        }
        
        if let birthday = birthday, birthday > Date() {
            errors.append("Birthday cannot be in the future")
        }
        
        if let weightKg = weightKg, weightKg <= 0 {
            errors.append("Weight must be positive")
        }
        
        return errors
    }
    
    enum CodingKeys: String, CodingKey {
        case name
        case species
        case breed
        case birthday
        case weightKg = "weight_kg"
        case activityLevel = "activity_level"
        case imageUrl = "image_url"
        case knownSensitivities = "known_sensitivities"
        case vetName = "vet_name"
        case vetPhone = "vet_phone"
    }
}

/// Pet update model for profile updates
struct PetUpdate: Codable {
    let name: String?
    let breed: String?
    let birthday: Date?
    let weightKg: Double?
    let activityLevel: PetActivityLevel?
    let imageUrl: String?
    let knownSensitivities: [String]?
    let vetName: String?
    let vetPhone: String?
    
    enum CodingKeys: String, CodingKey {
        case name
        case breed
        case birthday
        case weightKg = "weight_kg"
        case activityLevel = "activity_level"
        case imageUrl = "image_url"
        case knownSensitivities = "known_sensitivities"
        case vetName = "vet_name"
        case vetPhone = "vet_phone"
    }
}
